Add initial slew of tests for wide-layout attachment

squelart opened the pull request:
#### 0ea376b24d5bd5ed0d5649a940cc60fdf5862e5f
<pre>
Add initial slew of tests for wide-layout attachment
<a href="https://bugs.webkit.org/show_bug.cgi?id=254203">https://bugs.webkit.org/show_bug.cgi?id=254203</a>
rdar://106984208

Reviewed by Aditya Keerthi.

Most are copies of the non-&quot;wide&quot; tests.

Minor bugs were revealed and fixed:
- When setting/changing the top-level attachment&apos;s id or class (or even style) attribute, it was copied to the inner attachment, possibly modifying its rendering through css. So instead of indiscriminately copying every attribute, only the necessary ones related to the icon/thumbnail display are now copied.
- setFile function calls on the top-level attachment needed to be forwarded to the inner attachment to display the correct icon in some cases.

Render tree tests could become proper ref tests once the inner attachment is replaced with a standard HTML image element.

* LayoutTests/TestExpectations:
* LayoutTests/fast/attachment/cocoa/wide-attachment-class-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-class.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom-expected.txt: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-id-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-id.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-rendering.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event-expected.txt: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-wrapping-action-expected-mismatch.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-wrapping-action.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process-expected.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Added.
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
When first constructed, only relevant attributes are copied into the inner attachment.

(WebCore::HTMLAttachmentElement::setFile):
File details may be set through this function (instead of through element attributes), so the call must be duplicated into the inner attachment.

(WebCore::HTMLAttachmentElement::parseAttribute):
When attributes are modified, only the ones relevant to the attachment display are forwarded to the inner attachment.

Canonical link: <a href="https://commits.webkit.org/262067@main">https://commits.webkit.org/262067@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/42ac9694d68551962a0143c5c2698414576dfa5c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/439 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/453 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/472 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/439 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/391 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/438 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/495 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/521 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/637 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/445 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/428 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/425 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/473 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/469 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/414 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/463 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/382 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/431 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/412 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/426 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/108 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/420 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (commented):
The save event should be tested as well

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -211,7 +211,14 @@ void HTMLAttachmentElement::ensureModernShadowTree(ShadowRoot& root)
     m_innerLegacyAttachment = adoptRef(*new HTMLAttachmentElement(HTMLNames::attachmentTag, document()));
     m_innerLegacyAttachment->m_implementation = Implementation::ImageOnly;
-    m_innerLegacyAttachment->cloneAttributesFromElement(*this);
+    auto copyAttribute = [this](const QualifiedName& attr) {
+        m_innerLegacyAttachment->setAttributeWithoutSynchronization(attr, attributeWithoutSynchronization(attr));
+    };
+    copyAttribute(actionAttr);
+    copyAttribute(progressAttr);
+    copyAttribute(subtitleAttr());
+    copyAttribute(titleAttr);
+    copyAttribute(typeAttr);
```
The ChangeLog should explain the reason for this change (the observed issue and how this fixes it).

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -357,6 +364,8 @@ URL HTMLAttachmentElement::blobURL() const
 void HTMLAttachmentElement::setFile(RefPtr<File>&& file, UpdateDisplayAttributes updateAttributes)
 {
     m_file = WTFMove(file);
+    if (m_innerLegacyAttachment)
+        m_innerLegacyAttachment->setFile(m_file.copyRef(), updateAttributes);
```
The ChangeLog should explain the reason for this change (the observed issue and how this fixes it).

pxlcoder commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -433,9 +445,6 @@ void HTMLAttachmentElement::parseAttribute(const QualifiedName& name, const Atom
     } else if (name == saveAttr())
         updateSaveButton(!value.isNull());
 
-    if (m_innerLegacyAttachment)
-        m_innerLegacyAttachment->setAttributeWithoutSynchronization(name, value);
-
```
The ChangeLog should explain the reason for this change (the observed issue and how this fixes it).

pxlcoder commented on LayoutTests/fast/attachment/cocoa/wide-attachment-save-event.html:
```diff
@@ -0,0 +1,99 @@
+<!DOCTYPE html><!-- webkit-test-runner [ AttachmentElementEnabled=true AttachmentWideLayoutEnabled=true ] -->
+<html>
+<head>
+<script src="../../../resources/js-test-pre.js"></script>
```
New tests can just use `js-test.js` (no need to use -pre and -post).

squelart replied:
Actually, changing to `js-test.js` in this file makes the output worse, with the description appearing just before the first `await` in the async test function! (Instead of at the top.)

But I can use `js-test.js` in wide-attachment-dom.html.

pxlcoder reviewed (approved):
Note that the long term plan is to make these ref tests, but for now we are using render tree dumps, matching existing `<attachment>` tests.

squelart committed:
Add initial slew of tests for wide-layout attachment
https://bugs.webkit.org/show_bug.cgi?id=254203
rdar://106984208

Reviewed by Aditya Keerthi.

Most are copies of the non-"wide" tests.

Minor bugs were revealed and fixed:
- When setting/changing the top-level attachment's id or class (or even style) attribute, it was copied to the inner attachment, possibly modifying its rendering through css. So instead of indiscriminately copying every attribute, only the necessary ones related to the icon/thumbnail display are now copied.
- setFile function calls on the top-level attachment needed to be forwarded to the inner attachment to display the correct icon in some cases.

Render tree tests could become proper ref tests once the inner attachment is replaced with a standard HTML image element.

* LayoutTests/TestExpectations:
* LayoutTests/fast/attachment/cocoa/wide-attachment-class-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-class.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-default-icon.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom-expected.txt: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-dom.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-folder-icon.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-icon-from-file-extension.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-id-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-id.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-rendering.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event-expected.txt: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-save-event.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti-expected.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-uti.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-wrapping-action-expected-mismatch.html: Added.
* LayoutTests/fast/attachment/cocoa/wide-attachment-wrapping-action.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process-expected.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-element-gpu-process.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic-expected.txt: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-image-controls-basic.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute-expected.html: Added.
* LayoutTests/fast/attachment/mac/wide-attachment-type-attribute.html: Added.
* LayoutTests/platform/ios-wk2/TestExpectations:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Added.
* LayoutTests/platform/mac-wk2/TestExpectations:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt: Added.
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
When first constructed, only relevant attributes are copied into the inner attachment.

(WebCore::HTMLAttachmentElement::setFile):
File details may be set through this function (instead of through element attributes), so the call must be duplicated into the inner attachment.

(WebCore::HTMLAttachmentElement::parseAttribute):
When attributes are modified, only the ones relevant to the attachment display are forwarded to the inner attachment.

Canonical link: https://commits.webkit.org/262067@main
